app/controllers/paymentsAndCharges/PaymentsAndChargeDetailsController.scala

```diff
@@ -61,39 +64,62 @@ class PaymentsAndChargeDetailsController @Inject()(override val messagesApi: Mes
-                  case Some(schemeFs) =>
-                    renderer.render(
-                      "paymentsAndCharges/paymentsAndChargeDetails.njk",
-                      summaryListData(srn, startDate, schemeFs, schemeDetails.schemeName, index)
-                    )
-                      .map(Ok(_))
-                  case _ =>
-                    Logger.warn(s"No Payments and Charge details found for the selected charge reference ${chargeRefs(index.toInt)}")
-                    Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
-                }
-              } catch {
-                case _: IndexOutOfBoundsException =>
-                  Logger.warn(
-                    s"[paymentsAndCharges.PaymentsAndChargeDetailsController][IndexOutOfBoundsException]:" +
-                      s"index $index of collection length ${chargeRefs.length} attempted"
-                  )
+                  .groupAndSortByStartDate(schemeFS, startDate.getYear)
+
+              val chargeRefsGroupedAndSorted: Seq[(LocalDate, Seq[String])] =
+                schemeFSGroupedAndSorted.map(
+                  dateAndFs => {
+                    val (date, schemeFs) = dateAndFs
+                    (date, schemeFs.map(_.chargeReference))
+                  }
+                )
+
+              (
+                schemeFSGroupedAndSorted.find(_._1 == startDate),
+                chargeRefsGroupedAndSorted.find(_._1 == startDate)
+              ) match {
+                case (Some(dateSchemeFs), Some(dateChargeRefs)) =>
+                  val (_, seqSchemeFs) = dateSchemeFs
+                  val (_, seqChargeRefs) = dateChargeRefs
+                  try {
+                    seqSchemeFs.find(_.chargeReference == seqChargeRefs(index.toInt)) match {
+                      case Some(schemeFs) =>
+                        renderer.render(
+                          template = "paymentsAndCharges/paymentsAndChargeDetails.njk",
+                          ctx = summaryListData(srn, startDate, schemeFs, schemeDetails.schemeName, index)
+                        ).map(Ok(_))
+                      case _ =>
+                        Logger.warn(
+                          s"No Payments and Charge details found for the " +
+                            s"selected charge reference ${seqChargeRefs(index.toInt)}"
+                        )
+                        Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
+                    }
+                  } catch {
+                    case _: IndexOutOfBoundsException =>
+                      Logger.warn(
+                        s"[paymentsAndCharges.PaymentsAndChargeDetailsController][IndexOutOfBoundsException]:" +
+                          s"index $startDate/$index of attempted"
+                      )
+                      Future.successful(Redirect(controllers.routes.SessionExpiredController.onPageLoad()))
```

Comment:
maybe rather than repeating this several times just have a private def and call it?